### PR TITLE
Proposal to add Datastore Event Class 6005

### PR DESCRIPTION
**Pull request by pladam**
#### Description of changes: 

The following changes are required for this event class:

### New Objects
- database
- databucket
- table
- query_info

### CODE REVIEW

**Comment by floydtree** [comment]
Hey @Aniak5 would like to get your review on this one since you had created a similar class a few months back.

File: dictionary.json [hunk 1]
```diff
@@ -1057,6 +1057,87 @@
...
+        }
+      },
+      "type": "integer_t"
+    },
+    "databucket": {
+      "caption": "Databucket",
+      "description": "The databucket object represents a databucket is a basic container that holds data, typically organized through the use of data partitions.",
+      "type": "databucket"
+    },
+    "databucket_type": {
```
[the remark this answers is not in the document]
Comment (pladam) [reply dictionary.json]: Removed specific types and replaced with updated type util

File: dictionary.json [hunk 2]
```diff
@@ -1057,6 +1057,87 @@
...
+      "caption": "Database",
+      "description": "The database type.",
+      "type": "database"
+    },
+    "databucket_type_id": {
```
[the remark this answers is not in the document]
Comment (pladam) [reply dictionary.json]: Removed specific types and replaced with updated type util

File: dictionary.json [hunk 3]
```diff
@@ -1057,6 +1057,87 @@
...
+        "3": {
+          "caption": "GCP Bucket"
+        },
+        "99": {
+          "caption": "Other"
+        }
+      },
+      "type": "string_t"
+    },
+    "datastore_query": {
```
[the remark this answers is not in the document]
Comment (pladam) [reply dictionary.json]: updated to query_string

File: dictionary.json [hunk 4]
```diff
@@ -1889,6 +1970,11 @@
       "description": "The most recent detection time of the activity or object. See specific usage.",
       "type": "timestamp_t"
     },
+    "last_updated_time": {
```
[the remark this answers is not in the document]
Comment (pladam) [reply dictionary.json]: updated to modified time

File: dictionary.json [hunk 5]
```diff
@@ -2329,6 +2415,11 @@
       "description": "The EPSS score's percentile representing relative importance and ranking of the score in the larger EPSS dataset.",
       "type": "float_t"
     },
+    "partition": {
```
[the remark this answers is not in the document]
Comment (pladam) [reply dictionary.json]: remove, not needed within the schema at this time

File: events/application/datastore_activity.json [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            "requirement": "optional"
+        },
+        "databucket": {
+            "group": "primary",
+            "requirement": "optional"
+        },
+        "table": {
+            "group": "primary",
+            "requirement": "optional"
+        },
```
Comment (Aniak5) [inline events/application/datastore_activity.json:51]: Given that this is `datastore` activity, should we add an `at_least_one` constraint for `database`, `databucket`, or `table`?

Comment (pladam) [reply events/application/datastore_activity.json:51]: I think thats a good idea

**Review by zschmerber — DISMISSED** [review]
looks good to me

File: dictionary.json [hunk 7]
```diff
@@ -1057,6 +1057,16 @@
...
       "type": "json_t"
     },
+    "database": {
+      "caption": "Database",
+      "description": "The database object represents a database which contains an organized collection of data or a type of data.",
+      "type": "database"
+    },
+    "databucket": {
+      "caption": "Databucket",
+      "description": "The databucket object represents a databucket is a basic container that holds data, typically organized through the use of data partitions.",
```
Comment (Aniak5) [inline dictionary.json:1067]: We could make this a bit more clear by rewording to something like `The data bucket object encapsulates details about a container that stores data, typically organized using data partitions.`

File: dictionary.json [hunk 8]
```diff
@@ -1057,6 +1057,16 @@
       "description": "The additional data that is associated with the event or object. See specific usage.",
       "type": "json_t"
     },
+    "database": {
+      "caption": "Database",
+      "description": "The database object represents a database which contains an organized collection of data or a type of data.",
+      "type": "database"
+    },
+    "databucket": {
```
[the remark this answers is not in the document]
Comment (zschmerber) [reply dictionary.json:1065]: If we need to change it it would be good to avoid "_" as they often are used in conventions. So databucket or bucket would be my vote. 
https://github.com/ocsf/ocsf-docs/blob/main/Understanding%20OCSF.md#conventions

File: objects/query_info.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  "extends": "object",
+  "name": "query_info",
+  "attributes": {
+    "query_time": {
+      "description": "The time when the query was run.",
+      "requirement": "optional"
+    },
+    "query_string": {
+      "caption": "Query String",
+      "description": "A string representating the query code being run. For example: SELECT *</code>",
```
Comment (Aniak5) [inline objects/query_info.json:13]: Misspelled `representating` should be `representing`

File: objects/query_info.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+      "requirement": "optional"
```
Comment (Aniak5) [inline objects/query_info.json:14]: The `query_string` should probably be `required` within this object.

File: objects/query_info.json [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    "data": {
+      "description": "The data returned from the query execution.",
+      "requirement": "optional"
+    },
+    "bytes_out": {
+      "description": "The size of the data returned from the query.",
```
Comment (Aniak5) [inline objects/query_info.json:21]: I have mixed feelings about the `bytes_out` directionality. It makes sense from the perspective of the datastore because its the bytes leaving the data store but from the perspective of a query it could be thought of as `bytes_in`. What are your thoughts on just using `bytes` to make this object more generic?

File: dictionary.json [hunk 12]
```diff
@@ -2563,6 +2573,11 @@
       "description": "The Domain Name System (DNS) query.",
       "type": "dns_query"
     },
+    "query_info": {
+      "caption": "Query Info",
+      "description": "The query info object contains data relating to accessing data within a datastore. If you need to access, manipulate, delete, or retrieve data from a datastore, you'll need a database query written using a specific syntax.",
```
Comment (Aniak5) [inline dictionary.json:2578]: Maybe reword this a bit to something along the lines of: "The query info object holds information related to data access within a datastore. To access, manipulate, delete, or retrieve data from a datastore, a database query must be written using a specific syntax."

File: objects/database.json [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  "description": "The database object represents a database which contains an organized collection of data or a type of data.",
+  "extends": "object",
+  "name": "database",
+  "attributes": {
+    "created_time": {
+      "description": "The time when the database was known to have been created.",
+      "requirement": "optional"
+    },
+    "modified_time": {
+      "description": "The most recent detection time of the activity or object. See specific usage.",
```
Comment (Aniak5) [inline objects/database.json:12]: The description needs to be updated, could be: `The most recent time when any changes, updates, or modifications were made within the database.`

File: objects/database.json [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      "requirement": "optional"
+    },
+    "desc": {
+      "caption": "Description",
```
Comment (Aniak5) [inline objects/database.json:16]: Don't need a caption here as it is identical to the dictionary. Unless we want to change it to `Database Description`

File: objects/database.json [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+          "caption": "Other"
+        }
+      }
+    },
+    "name": {
+      "description": "The database name, ordinarily as assigned by a database administrator."
+    },
+    "uid": {
+      "description": "The unique identifier of the database."
+    }
```
Comment (Aniak5) [inline objects/database.json:70]: Something in this object should be required so that we don't wind up with an empty `database` struct. Maybe a `at_least_one` constraint with `name` and `uid`.

File: objects/databucket.json [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  "description": "The databucket object represents a databucket is a basic container that holds data, typically organized through the use of data partitions.",
+  "extends": "object",
+  "name": "databucket",
+  "attributes": {
+    "created_time": {
+      "description": "The time when the databucket was known to have been created.",
+      "requirement": "optional"
+    },
+    "modified_time": {
+      "description": "The most recent detection time of the activity or object. See specific usage.",
```
Comment (Aniak5) [inline objects/databucket.json:12]: The description needs to be updated, could be: `The most recent time when any modifications were made to the databucket`.

File: objects/table.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  "description": "The table object represents a table within a structured relational database or datastore, which contains columns and rows of data that are able to be create, updated, deleted and queried.",
+  "extends": "object",
+  "name": "table",
+  "attributes": {
+    "created_time": {
+      "description": "The time when the table was known to have been created.",
+      "requirement": "optional"
+    },
+    "modified_time": {
+      "description": "The most recent detection time of the activity or object. See specific usage.",
```
Comment (Aniak5) [inline objects/table.json:12]: The description needs to be updated, could be: `The most recent time when any modifications were made to the table`.

File: objects/query_info.json (hunk 9 above)
Comment (adplotzk) [reply objects/query_info.json]: updated.

File: objects/query_info.json (hunk 10 above)
Comment (adplotzk) [reply objects/query_info.json]: updated.

File: events/application/datastore_activity.json (hunk 6 above)
Comment (adplotzk) [reply events/application/datastore_activity.json:51]: updated

**Review by adplotzk — DISMISSED** [review]

**Review by mikeradka — DISMISSED** [review]

**Review by mikeradka — APPROVED** [review]

**Review by pagbabian-splunk — DISMISSED** [review]

File: objects/databucket.json [hunk 18]
```diff
@@ -0,0 +1,64 @@
+{
+  "caption": "Databucket",
+  "description": "The databucket object represents a databucket is a basic container that holds data, typically organized through the use of data partitions.",
+  "extends": "object",
```
[the remark this answers is not in the document]
Comment (floydtree) [reply objects/databucket.json:4]: `.. used for databuckets which are typically cloud-based containers..` This is a bit misleading, since containers are a thing by themselves. Good to keep it simple the way Paul suggested above.

**Review by floydtree — CHANGES_REQUESTED** [review]

File: objects/databucket.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (pladam) [reply objects/databucket.json:4]: Updated based on specifications above

**Review by pladam — COMMENTED** [review]
Updated description. Please review

File: objects/databucket.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (floydtree) [reply objects/databucket.json:4]: Thanks, but same applies to the description in the dictionary. You'll need to update that as well.

File: events/application/datastore_activity.json [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+                },
+                "7": {
+                    "caption": "Delete",
+                    "description": "The datastore activity in the event pertains to a 'Delete' operation."
+                }
+            }
+        },
+        "database": {
+            "group": "primary",
+            "requirement": "optional"
```
Comment (floydtree) [inline events/application/datastore_activity.json:42]: Better, to mark these - database, databucket as recommended, considering these are the key attributes in this class. Typically, all the primary attributes should be recommended.

File: events/application/datastore_activity.json [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+        },
+        "table": {
+            "group": "primary",
+            "requirement": "optional"
+        },
+        "query_info": {
+            "group": "primary",
+            "requirement": "optional"
+        },
+        "dst_endpoint": {
```
Comment (floydtree) [inline events/application/datastore_activity.json:56]: Good to act a specific description(considering you have added one for src_endpoint). It is particularly important in here since, there are quite a few similar objects (should a mapper populate dst_endpoint if they are populating databucket? for instance) A description clarifying these relationships will help.

File: objects/database.json [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+      "description": "The size of database, in bytes.",
+      "requirement": "optional"
+    },
+    "groups": {
+      "description": "The group names to which the database belongs.",
+      "requirement": "optional"
+    },
+    "type": {
+      "description": "The database type.",
+      "requirement": "required"
```
Comment (floydtree) [inline objects/database.json:29]: The string siblings should be recommended (current strategy used throughout OCSF). Unless there is a good reason to enforce people to populate both, but that would be a bigger discussion and not limited to this object.

File: objects/databucket.json [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+      "description": "A file within a databucket.",
+      "requirement": "optional"
+    },
+    "groups": {
+      "description": "The group names to which the databucket belongs.",
+      "requirement": "optional"
+    },
+    "type": {
+      "description": "The database type.",
+      "requirement": "required"
```
Comment (floydtree) [inline objects/databucket.json:34]: Same comment as in the database object - String siblings -recommended.

File: objects/table.json [hunk 23]
```diff
@@ -0,0 +1,35 @@
+{
+  "caption": "Table",
+  "description": "The table object represents a table within a structured relational database or datastore, which contains columns and rows of data that are able to be create, updated, deleted and queried.",
+  "extends": "object",
```
Comment (floydtree) [inline objects/table.json:4]: extend _entity instead, to inherit the constraints for uid, name.

File: objects/query_info.json [hunk 24]
```diff
@@ -0,0 +1,31 @@
+{
+  "caption": "Query Information",
+  "description": "The query info object contains data relating to accessing data within a datastore. If you need to access, manipulate, delete, or retrieve data from a datastore, you'll need a database query written using a specific syntax.",
+  "extends": "object",
```
Comment (floydtree) [inline objects/query_info.json:4]: extend _entity instead, to inherit the constraints for uid, name.

File: objects/databucket.json [hunk 25]
```diff
@@ -0,0 +1,64 @@
+{
+  "caption": "Databucket",
+  "description": "The data bucket object is a basic container that holds data, typically organized through the use of data partitions.",
+  "extends": "object",
```
Comment (floydtree) [inline objects/databucket.json:4]: extend _entity instead, to inherit the constraints for uid, name.

File: objects/database.json [hunk 26]
```diff
@@ -0,0 +1,71 @@
+{
+  "caption": "Database",
+  "description": "The database object is used for databases which are typically datastore services that contain an organized collection of structured and unstructured data or a types of data.",
+  "extends": "object",
```
Comment (floydtree) [inline objects/database.json:4]: extend _entity instead, to inherit the constraints for uid, name.

**Review by floydtree — CHANGES_REQUESTED** [review]
Good stuff! The class looks almost ready, however take a look at the specific comments for the changes that are still needed.

Also, please ensure the commit messages are self explanatory, the current commit messages do not convey what is being changed. This is crucial if we ever to have to go back and look for specific commits, or creating a release.

File: events/application/datastore_activity.json (hunk 19 above)
Comment (adplotzk) [reply events/application/datastore_activity.json:42]: changing to recommended, also I have these at "at least one" required

File: events/application/datastore_activity.json [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+        "http_request": {
+            "description":"Details about the underlying http request.",
+            "group": "primary",
+            "requirement": "optional"
+        },
+        "actor": {
+            "group": "primary",
+            "requirement": "required"
+        },
+        "resources": {
```
[the remark this answers is not in the document]
Comment (adplotzk) [reply events/application/datastore_activity.json]: I agree this is not needed in this context

File: objects/databucket.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (adplotzk) [reply objects/databucket.json]: updated.

File: objects/table.json (hunk 23 above)
Comment (adplotzk) [reply objects/table.json]: updated

File: objects/database.json (hunk 21 above)
Comment (adplotzk) [reply objects/database.json]: changing type to recommended

File: objects/databucket.json (hunk 22 above)
Comment (adplotzk) [reply objects/databucket.json]: changing type to recommended

File: dictionary.json [hunk 28]
```diff
@@ -1057,6 +1057,16 @@
...
       "type": "json_t"
     },
+    "database": {
+      "caption": "Database",
+      "description": "The database object is used for databases which are typically datastore services that contain an organized collection of structured and unstructured data or a types of data.",
+      "type": "database"
+    },
+    "databucket": {
+      "caption": "Databucket",
+      "description": "The databucket object is used for databuckets which are typically cloud-based container services used for object storage. Databuckets are often organized in directory like structures called partitions.",
```
[the remark this answers is not in the document]
Comment (adplotzk) [reply dictionary.json]: I've addressed all the comments here - please review

**Review by zschmerber — APPROVED** [review]
Thanks for all the hard work on this.

**Comment by pagbabian-splunk** [comment]
Looks good now for DML activities (data manipulation).  Would there be events associated with DML activities (i.e. metadata create, update, delete; create table, update table, delete table)?

File: events/application/datastore_activity.json [hunk 29]
```diff
@@ -0,0 +1,83 @@
+{
+    "uid": 5,
+    "description": "Datastore events describe general activities (Read, Update, Query, Delete, etc.) which affect datastores or data within those datastores, e.g. (AWS RDS, AWS S3)",
```
Comment (Aniak5) [inline events/application/datastore_activity.json:3]: Nit: Can we add a `.` at the end of this description?

File: objects/database.json [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+    "modified_time": {
+      "description": "The most recent time when any changes, updates, or modifications were made within the database.",
+      "requirement": "optional"
+    },
+    "desc": {
+      "caption": "Description",
+      "requirement": "optional"
+    },
+    "size": {
+      "description": "The size of database, in bytes.",
```
Comment (Aniak5) [inline objects/database.json:20]: Nit: In its current form it feels like the sentence is incomplete, can we change it to `The size of the database in bytes.`

File: objects/databucket.json [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+      "description": "The most recent time when any changes, updates, or modifications were made within the databucket.",
+      "requirement": "optional"
+    },
+    "desc": {
+      "caption": "Description",
+      "description": "The description of the databucket.",
+      "requirement": "optional"
+    },
+    "size": {
+      "description": "The size of data bucket, in bytes.",
```
Comment (Aniak5) [inline objects/databucket.json:21]: Nit: Same comment as above and for consistency lets stick with `databucket` instead of `data bucket`. `The size of the the databucket in bytes.`.

File: objects/databucket.json [hunk 32]
```diff
@@ -0,0 +1,64 @@
...
+    "file": {
+      "description": "A file within a databucket.",
+      "requirement": "optional"
+    },
+    "groups": {
+      "description": "The group names to which the databucket belongs.",
+      "requirement": "optional"
+    },
+    "type": {
+      "description": "The database type.",
```
Comment (Aniak5) [inline objects/databucket.json:33]: This should be `The databucket type.`

File: objects/query_info.json [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+  "extends": "_entity",
+  "name": "query_info",
+  "attributes": {
+    "query_time": {
+      "description": "The time when the query was run.",
+      "requirement": "optional"
+    },
+    "query_string": {
+      "caption": "Query String",
+      "description": "A string representing the query code being run. For example: SELECT *</code>",
```
Comment (Aniak5) [inline objects/query_info.json:13]: Missing the `<code>` before the `SELECT`. The example might be a little more clear if it is changed to `SELECT * FROM my_table` but i'm fine w/it as is too.

File: events/application/datastore_activity.json (hunk 29 above)
Comment (adplotzk) [reply events/application/datastore_activity.json]: updated

File: objects/database.json (hunk 30 above)
Comment (adplotzk) [reply objects/database.json]: updated

File: objects/databucket.json (hunk 31 above)
Comment (adplotzk) [reply objects/databucket.json]: updates

File: objects/databucket.json (hunk 32 above)
Comment (adplotzk) [reply objects/databucket.json]: updated

File: objects/query_info.json (hunk 33 above)
Comment (adplotzk) [reply objects/query_info.json]: updated

**Review by Aniak5 — DISMISSED** [review]
Looks awesome @adplotzk, excited to finally have this event class!!

File: objects/table.json [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+      "description": "The most recent time when any changes, updates, or modifications were made within the table.",
+      "requirement": "optional"
+    },
+    "desc": {
+      "caption": "Description",
+      "description": "The description of the table.",
+      "requirement": "optional"
+    },
+    "size": {
+      "description": "The size of data table, in bytes.",
```
Comment (Aniak5) [inline objects/table.json:21]: One last change! `The size of the data table in bytes.`

Comment (pladam) [reply objects/table.json]: updated

**Review by Aniak5 — APPROVED** [review]

**Review by zschmerber — APPROVED** [review]

**Review by floydtree — APPROVED** [review]
